### ISSUE TITLE
Almost any interface change in the Fluid repo results in IFluidObject typing incompatibility

### ISSUE THREAD

**Issue by leeviana** [microsoft/FluidFramework#2566]
## Describe the bug
Anything that references an IFluidObject type from different versions are pretty much always guaranteed to be typescript incompatible in a "major version" sort of way because of the number of "provides" contracts on IFluidObject . This means we can pretty much never have two different versions of packages on either side of an IFluidObject exchange, regardless of whether you're even using any interfaces that have actually changed. For instance, if somebody makes a random change to IFluidDataStoreContext, through a really deep dependency chain involving views and IHostRuntime, this essentially means a major bump in IFluidObject/core-interfaces.

### To Reproduce
Make a change to almost any interface in the Fluid repo. Likely the IFluidObject interface in the version of the code published is not compatible with the version of code right before.

### Expected behavior
IFluidObject interfaces should be somewhat durable across Framework changes, especially in interfaces that aren't being immediately used.

**Comment by DLehenbauer** [comment]
Found an example of 'IFluidDataStoreFactory' being incompatible with 'IFluidDataStoreFactory', where both packages are pulling v0.26.1 from npmjs:
```
ERROR in /workspaces/ps/danlehen/graphql/client/src/containerCode.ts
[tsl] ERROR in /workspaces/ps/danlehen/graphql/client/src/containerCode.ts(21,5)
      TS2345: Argument of type 'Map<string, Promise<Readonly<Partial<IProvideFluidDataStoreRegistry & IProvideFluidDataStoreFactory>>>>' is not assignable to parameter of type 'NamedFluidDataStoreRegistryEntries'.
  The types returned by '[Symbol.iterator]().next(...)' are incompatible between these types.
    Type 'IteratorResult<[string, Promise<Readonly<Partial<IProvideFluidDataStoreRegistry & IProvideFluidDataStoreFactory>>>], any>' is not assignable to type 'IteratorResult<NamedFluidDataStoreRegistryEntry, any>'.
      Type 'IteratorYieldResult<[string, Promise<Readonly<Partial<IProvideFluidDataStoreRegistry & IProvideFluidDataStoreFactory>>>]>' is not assignable to type 'IteratorResult<NamedFluidDataStoreRegistryEntry, any>'.
        Type 'IteratorYieldResult<[string, Promise<Readonly<Partial<IProvideFluidDataStoreRegistry & IProvideFluidDataStoreFactory>>>]>' is not assignable to type 'IteratorYieldResult<NamedFluidDataStoreRegistryEntry>'.
          Type '[string, Promise<Readonly<Partial<IProvideFluidDataStoreRegistry & IProvideFluidDataStoreFactory>>>]' is not assignable to type 'NamedFluidDataStoreRegistryEntry'.
            Type 'Promise<Readonly<Partial<import("/workspaces/ps/danlehen/graphql/common/node_modules/@fluidframework/runtime-definitions/dist/dataStoreRegistry").IProvideFluidDataStoreRegistry & import("/workspaces/ps/danlehen/graphql/common/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IProvideFluidDataS...' is not assignable to type 'Promise<Readonly<Partial<import("/workspaces/ps/danlehen/graphql/client/node_modules/@fluidframework/runtime-definitions/dist/dataStoreRegistry").IProvideFluidDataStoreRegistry & import("/workspaces/ps/danlehen/graphql/client/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IProvideFluidDataS...'.
              Type 'Readonly<Partial<import("/workspaces/ps/danlehen/graphql/common/node_modules/@fluidframework/runtime-definitions/dist/dataStoreRegistry").IProvideFluidDataStoreRegistry & import("/workspaces/ps/danlehen/graphql/common/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IProvideFluidDataStoreFact...' is not assignable to type 'Readonly<Partial<import("/workspaces/ps/danlehen/graphql/client/node_modules/@fluidframework/runtime-definitions/dist/dataStoreRegistry").IProvideFluidDataStoreRegistry & import("/workspaces/ps/danlehen/graphql/client/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IProvideFluidDataStoreFact...'.
                Types of property 'IFluidDataStoreFactory' are incompatible.
                  Type 'import("/workspaces/ps/danlehen/graphql/common/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IFluidDataStoreFactory | undefined' is not assignable to type 'import("/workspaces/ps/danlehen/graphql/client/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IFluidDataStoreFactory | undefined'.
                    Type 'import("/workspaces/ps/danlehen/graphql/common/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IFluidDataStoreFactory' is not assignable to type 'import("/workspaces/ps/danlehen/graphql/client/node_modules/@fluidframework/runtime-definitions/dist/dataStoreFactory").IFluidDataStoreFactory'.
```
Fixed by casting to 'any' here:
```ts
export const GraphContainerRuntimeFactory = new ContainerRuntimeFactoryWithDefaultDataStore(
    GraphInstantiationFactory.type,
    new Map([
        GraphInstantiationFactory.registryEntry as any,   // <-- Fixed by 'as any'
    ]),
);
```

**Comment by curtisman** [comment]
@DLehenbauer what are the version of @fluidframework/runtime-definitions in `/workspaces/ps/danlehen/graphql/common` vs `/workspaces/ps/danlehen/graphql/client`

**Comment by leeviana** [comment]
Yeah, unfortunately things like this have led to a lot of similar 'any' casting where we test different versions of host/loader etc. :(

**Comment by DLehenbauer** [comment]
@curtisman - It's '0.26.1' in both cases.  I did a recursive diff of all the fluidframework packages like this:
```
diff -r /workspaces/ps/danlehen/graphql/common/node_modules/@fluidframework /workspaces/ps/danlehen/graphql/client/node_modules/@fluidframework
```
...and they are 100% identical (except for the "_where:" field generated by 'npm i' in package.json...  which differs as expected).

(I'm still poking at it a bit.)

**Comment by leeviana** [comment]
This is still an issue, what is the standard here, just to leave a comment to prevent if from being closed?

**Comment by markfields** [comment]
Still interested in this but about to go on leave so removing my assignment.

**Comment by Stevenic** [comment]
FYI... We ran into a similar issue with the Bot Framework SDK's. The lesson we learned is that you should always try to favor classes over interfaces as interfaces are essentially immutable and can almost never be updated. The advantage of classes is that you can easily extend them in future releases and provide default behaviors so that you don't break compatibility with previous versions.

**Comment by anthony-murphy** [comment]
@Stevenic Thanks for this! curious if you had cases where consumer would be the implementor? did you end up going with abstract classes is these cases?

**Comment by anthony-murphy** [comment]
will be fixed by #7613
